rasterio/transform.py

```diff
@@ -175,16 +170,15 @@ def xy(transform, rows, cols, offset='center'):
 
     xs = []
     ys = []
-    for col, row in zip(cols, rows):
-        x, y = transform * transform.translation(coff, roff) * (col, row)
+    T = transform * transform.translation(coff, roff)
+    for pt in zip(cols, rows):
+        x, y = T * pt
```

Comment:
```suggestion
    xs, ys = zip(*[T * pt for pt in zip(cols, rows)])
```

More efficient than looping or appending. Using numpy to broadcast would probably be even faster.

Reply:
In one of my experiments, I test the essence of this concept, though I didn't do the comprehension. I wasn't sure how far I could go with changing things before I crossed a boundary.
I tried:
```python
locs = []
T = ...
for pt in zip(cols, rows):
    locs.append(T * pt)
xs, ys = tuple(zip(*locs))
```
which didn't really give much of a speed up over the two list method. 

If we make that a generator expression, then it is _slightly_ faster.
`xs, ys = zip(*(T * pt for pt in zip(cols, rows)))`

I wasn't sure what the stance was on numpy. If we did a numpy version, it would be **significantly** faster for multiple points (https://github.com/mapbox/rasterio/issues/2108)

Reply:
Aha! I overlooked that one. Rasterio depends on numpy all the way through, so it's fine to use it within the transform function.

Reply:
What was confusing me was this comment: https://github.com/mapbox/rasterio/blob/master/rasterio/dtypes.py#L3-L5
and this statement on affine: https://github.com/sgillies/affine/issues/44
Those combined with the local imports of numpy in dtypes.py left me wondering if there was something particular about using numpy even thought the project clearly depends on numpy elsewhere.